src/junction/markdown/checklists.py

```diff
@@ -0,0 +1,65 @@
+import re
+from typing import List, Any
+
+from markdown import Markdown
+from markdown.extensions import Extension
+from markdown.postprocessors import Postprocessor
+
+
+class ChecklistExtension(Extension):
+    def extendMarkdown(self, md: Markdown) -> None:
+        postprocessor = ChecklistPostprocessor(md)
+        md.postprocessors.add('checklist', postprocessor, '>raw_html')
+
+
+class ChecklistPostprocessor(Postprocessor):
+    """Markdown extension that supports GitHub styled checklists for tasks, based on https://github.com/FND/markdown-checklist
+
+    For example:
+    - [ ] An incomplete task
+    - [x] A completed task
+    """
+
+    item_with_children_pattern = re.compile(r'<li>\[([ Xx])\]((?!</li>).*)<ul>', re.MULTILINE)
+    item_paragraph_pattern = re.compile(r'<p>\[([ Xx])\](.*)</p>', re.MULTILINE)
+    item_pattern = re.compile(r'<li>\[([ Xx])\](.*)</li>', re.MULTILINE)
+
+    def run(self, html: str) -> str:
+        # Converts nested lists that start with a task
+        html = re.sub(self.item_with_children_pattern, self._convert_item_with_children, html)
+        # Converts paragraphs with the task syntax, in case the HTML parser created an extra line
+        html = re.sub(self.item_paragraph_pattern, self._convert_item, html)
+        # Converts regular list items
+        html = re.sub(self.item_pattern, self._convert_item, html)
+        return html
+
+    def _convert_item(self, match: re.Match) -> str:
+        """
+        Converts an item without children, uses the offset of the match from the start of the HTML
+        as the task ID
+        """
+        state, caption = match.groups()
+        return render_item(caption, state != ' ', match.start())
+
+    def _convert_item_with_children(self, match: re.Match) -> str:
+        """
+        Converts an item with children, uses the offset of the match from the start of the HTML
+        as the task ID
+        """
+        state, caption = match.groups()
+        return render_item(caption, state != ' ', match.start()) + "<ul><li>"
+
+
+def render_item(caption: str, checked: bool, task_id: int) -> str:
```

Comment:
IIUC, this means that each task in a list of multiple tasks ends up as a separate `<ac:task-list>`?  That's probably fine (and you said you tested it and it worked for your case), but I'm guessing "native" Confluence produces all adjacent tasks under a single task list.

The flip side is you would need to verify that every item in the current list level is formatted like a task before applying this transformation (or assume that 1st item is a task == all items should be treated like tasks.

I'm not sure how flexible Confluence normally is when it comes to mixing ul/ol/tasks and sublists so I may be leading you down an invalid path...take all of this with a grain of salt.

Reply:
You're right, it currently creates a new list for each task. Parsing a continued list of tasks would have been considerably more difficult. I haven't noticed any visual or functional side-effects in Confluence with the files I've tested.

I often mix tasks and regular bullet points in my own notes to mark sth that I need to do, versus something that I just noted down. So making everything a task wouldn't have helped in my own usecase.

Confluence seems to dislike creating tasks in nested lists, it just breaks out of the list when you mark sth as a task in the WYSIWYG editor. But it also parses and renders the nested lists and `ac:task` markup I've thrown at it.

I guess a good improvement would be to keep track of consecutive task items on the same indentation and create one `ac:task-list` for each.